docs/architecture/adr-001-redesign-knuu-async-approach.md

```diff
@@ -0,0 +1,55 @@
+# ADR 001: Redesigning Knuu for Asynchronous Testing Approach
+
+## Status
+
+Draft
+
+## Context
+
+The existing design of `Knuu` requires exposing the addresses and ports of the target instance running alongside BitTwister, complicating the configuration process. This setup requires Knuu to orchestrate the instances, configure them, and execute the test, leading to complexities in managing traffic shaping. Moreover, as the current model relies on port forwarding, scalability issues have surfaced, causing inconsistent errors and hindering seamless test execution.
+
+## Problem
+
+The current synchronous model necessitates exposing the target instance's address and ports to BitTwister via a sidecar. This centralized approach introduces dependencies and complexities in configuring the test environment. Additionally, the reliance on a proxy approach or port forwarding poses scalability issues, resulting in inconsistent errors during test execution.
+
+## Proposed Solution
+
+### A Distributed Approach with a Playbook Model
+
+Embrace a distributed playbook model similar to Testground, allowing asynchronous orchestration and execution of test scenarios:
```

Comment:
What will be the differences in use case between KNUU and Testground if we move in this direction? 

I think something that would be helpful for this ADR, and any large architecture overhaul, is to include some examples of desired test scenarios that we want to be able to support.

Reply:
Well in that case Knuu will be going towards the same direction of TG and we have to see which one to choose to work on. 
Knuu is built in house and needs a lot of love.
TG seems to be used by more than one team.

That's right we need to have some good examples. @Bidon15 can you please shed some lights here?

Reply:
Ok here is an example that I could come up with:

Let's say we have the following node Configuration:

- **150** Validators
- **300** Consensus Full Nodes
- **1000** Full Data Availability (DA) Nodes
- **100,000** Light DA Nodes

Playbook setup:
- **After reaching height 10**:
    - Adjust the bandwidth of 15% of Validators to 10Mbps.
    - Introduce a 50ms latency to 30% of Light DA Nodes.
    - Add a 20ms jitter to 25% of Full DA Nodes.

- **Test Duration**:
    - Keep the test running until height 10,000 
    - or 30 minutes, ensuring comprehensive evaluation.

- **Metrics collection**
    - Configured traffic shape on each node
    - Error logs of each node
    - Time to complete the sync of DA nodes
    - Number of missing blocks to sign
    - Number of missing blocks to propose
    - ...


Each worker node receives this playbook and based on that decides what to do.
The entire playbook could be written in a golang code just like TG.

Reply:
if we go either way of improving Knuu or work on TG, for both we will need heavy refactor. or rather building a significant portion of it. Each of them has it's own potentials and it is a bit hard to chose either ways. The structure that is proposed in this ADR is more align with TG; however, TG has a number of significant limitation that need to be resolved to be able to handle a large scale test.

@Bidon15 has worked a lot with TG, tell us what do you think?

Reply:
ah ok makes sense. 
So we could have a future where we just use knuu and testground is dead to us. 

Makes sense. 

Let's add the example that you listed here to the ADR as a reference.